test/segment/segmentTemplate.test.js

```diff
@@ -688,15 +688,463 @@ function(assert) {
+        uri: ''
+      },
+      number: 2,
+      resolvedUri: 'http://www.example.com/n-2.m4s',
+      timeline: 1,
+      uri: 'n-2.m4s'
+    }, {
+      duration: 2,
+      map: {
+        resolvedUri: 'http://www.example.com/',
+        uri: ''
+      },
+      number: 3,
+      resolvedUri: 'http://www.example.com/n-3.m4s',
+      timeline: 1,
+      uri: 'n-3.m4s'
+    }],
+    'segments correct with @availabilityStartTime set'
+  );
+
+  assert.deepEqual(
+    segmentsFromTemplate(Object.assign({}, basicAttributes,
+      { availabilityStartTime: 2, start: 4 }), []),
+    [{
+      duration: 2,
+      map: {
+        resolvedUri: 'http://www.example.com/',
+        uri: ''
+      },
+      number: 1,
+      resolvedUri: 'http://www.example.com/n-1.m4s',
+      timeline: 1,
+      uri: 'n-1.m4s'
+    }, {
+      duration: 2,
+      map: {
+        resolvedUri: 'http://www.example.com/',
+        uri: ''
+      },
+      number: 2,
+      resolvedUri: 'http://www.example.com/n-2.m4s',
+      timeline: 1,
+      uri: 'n-2.m4s'
+    }],
+   'segments correct with @availabilityStartTime and @start set');
+
+  assert.deepEqual(
+    segmentsFromTemplate(Object.assign({}, basicAttributes,
+      { timeShiftBufferDepth: 4 }, [])),
+    [{
+      duration: 2,
+      map: {
+        resolvedUri: 'http://www.example.com/',
+        uri: ''
+      },
+      number: 4,
+      resolvedUri: 'http://www.example.com/n-4.m4s',
+      timeline: 1,
+      uri: 'n-4.m4s'
+    }, {
+      duration: 2,
+      map: {
+        resolvedUri: 'http://www.example.com/',
+        uri: ''
+      },
+      number: 5,
+      resolvedUri: 'http://www.example.com/n-5.m4s',
+      timeline: 1,
+      uri: 'n-5.m4s'
+    }],
+   'segments correct with @timeShiftBufferDepth set');
+
+  assert.deepEqual(
+    segmentsFromTemplate(Object.assign({}, basicAttributes,
+      { clientOffset: -2000 }, [])),
+    [{
+      duration: 2,
+      map: {
+        resolvedUri: 'http://www.example.com/',
+        uri: ''
+      },
+      number: 1,
+      resolvedUri: 'http://www.example.com/n-1.m4s',
+      timeline: 1,
+      uri: 'n-1.m4s'
+    }, {
+      duration: 2,
+      map: {
+        resolvedUri: 'http://www.example.com/',
+        uri: ''
+      },
+      number: 2,
+      resolvedUri: 'http://www.example.com/n-2.m4s',
+      timeline: 1,
+      uri: 'n-2.m4s'
+    }, {
+      duration: 2,
+      map: {
+        resolvedUri: 'http://www.example.com/',
+        uri: ''
+      },
+      number: 3,
+      resolvedUri: 'http://www.example.com/n-3.m4s',
+      timeline: 1,
+      uri: 'n-3.m4s'
+    }, {
+      duration: 2,
+      map: {
+        resolvedUri: 'http://www.example.com/',
+        uri: ''
+      },
+      number: 4,
+      resolvedUri: 'http://www.example.com/n-4.m4s',
+      timeline: 1,
+      uri: 'n-4.m4s'
+    }],
+    'segments correct with given clientOffset');
+});
+
+QUnit.test('correctly handles duration with segmentTimeline', function(assert) {
```

Comment:
I think it would be good to add a scenario where the timeline is shifted and a `timescale !== 1` (you already have the time shifted example, but with timescale 1)